File: python-algorithm/leetcode/problem_649.py
"""649. Dota2 Senate
https://leetcode.com/problems/dota2-senate/
"""
from typing import List
import bisect


class Solution:
    def predict_party_victory(self, senate: str) -> str:
        def disable(idx_list: List[int], start: int):
            idx = bisect.bisect_right(idx_list, start)
            for j in range(idx, len(idx_list)):
                if can_vote[idx_list[j]]:
                    can_vote[idx_list[j]] = False
                    del idx_list[j]
                    return
            for j in range(idx):
                if can_vote[idx_list[j]]:
                    can_vote[idx_list[j]] = False
                    del idx_list[j]
                    return

        n = len(senate)
        can_vote = [True] * n
        r_list, d_list = [], []
        for i in range(n):
            (r_list if senate[i] == 'R' else d_list).append(i)
        while True:
            for i in range(n):
                if not can_vote[i]:
                    continue
                disable(d_list if senate[i] == 'R' else r_list, i)
                if not r_list:
                    return 'Dire'
                if not d_list:
                    return 'Radiant'

File: java-algo/src/main/java/com/leetcode/Problem649.java
package com.leetcode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 649. Dota2 Senate
 * https://leetcode.com/problems/dota2-senate/
 */
public class Problem649 {

    public String predictPartyVictory(String senate) {
        int n = senate.length();
        boolean[] disabled = new boolean[n];
        List<Integer> dires = new ArrayList<>(), radiants = new ArrayList<>();
        for (int i = 0; i < n; i++)
            (senate.charAt(i) == 'R' ? radiants : dires).add(i);
        while (true) {
            for (int i = 0; i < n; i++) {
                if (disabled[i]) continue;
                disable(senate.charAt(i) == 'R' ? dires : radiants, i, disabled);
                if (dires.size() == 0) return "Radiant";
                if (radiants.size() == 0) return "Dire";
            }
        }
    }

    private void disable(List<Integer> list, int start, boolean[] disabled) {
        int idx = Math.abs(Collections.binarySearch(list, start) + 1);
        for (int i = idx; i < list.size(); i++) {
            int pos = list.get(i);
            if (!disabled[pos]) {
                disabled[pos] = true;
                list.remove(i);
                return;
            }
        }
        for (int i = 0; i < idx; i++) {
            int pos = list.get(i);
            if (!disabled[pos]) {
                disabled[pos] = true;
                list.remove(i);
                return;
            }
        }
    }
}
